File: src/api/private/authenticate.ts
import { RequestFN } from "~/core/request-function";
import { AccessDeniedError, AccountDisabledError, AuthenticateError, BadCredentialsError, type SessionHandle } from "~/models";

export const authenticate = async (session: SessionHandle, challenge: string) => {
  const request = new RequestFN(session, "Authentification", {
    donnees: {
      connexion: 0, // NOTE: Probably the `accessKind` property, not sure though.
      challenge,
      espace: session.information.accountKind
    }
  });

  const response = await request.send();
  const data = response.data.donnees;

  // Handle potential errors.
  if (typeof data.Acces === "number" && data.Acces !== 0) {
    switch (data.Acces) {
      case 1: // EGenreErreurAcces.Identification
        throw new BadCredentialsError();

      case 2: // EGenreErreurAcces.Autorisation
      case 3: // EGenreErreurAcces.ConnexionClasse
      case 4: // EGenreErreurAcces.AucuneRessource
      case 5: // EGenreErreurAcces.Connexion
      case 7: // EGenreErreurAcces.FonctionAccompagnant
      case 8: // EGenreErreurAcces.AccompagnantAucunEleve
        throw new AccessDeniedError();

      case 6: // EGenreErreurAcces.BloqueeEleve
      case 10: // EGenreErreurAcces.CompteDesactive
        throw new AccountDisabledError();

      case 9: // EGenreErreurAcces.Message
        if (typeof data.AccesMessage !== "undefined") {
          let error: string = data.AccesMessage.message ?? "(none)";

          if (data.AccesMessage.titre) {
            error += `${data.AccesMessage.titre} ${error}`;
          }

          throw new AuthenticateError(error);
        }
    }
  }

  return data;
};

File: library/src/commonMain/kotlin/api/private/authenticate.kt
package api.private

import core.RequestFN
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*
import models.SessionInformation
import models.errors.AccessDeniedError
import models.errors.AuthenticateError
import models.errors.BadCredentialsError

suspend fun authenticate (sessionInfo: SessionInformation, challenge: String): JsonObject {
    val requestData = buildJsonObject {
        putJsonObject("donnees") {
            put("connexion", 0)
            put("challenge", challenge)
            put("espace", sessionInfo.accountKind.code)
        }
    }

    val request = RequestFN(sessionInfo, "Authentification", Json.encodeToString(requestData))
    val response = request.send()
    val data = Json.parseToJsonElement(response.data).jsonObject["donnees"]!!.jsonObject

    if (data["Acces"] != null)
        when (data["Acces"]!!.jsonPrimitive.int) {
            1 -> throw BadCredentialsError()
            2 -> {/*EGenreErreurAcces.Autorisation*/}
            3 -> {/*EGenreErreurAcces.ConnexionClasse*/}
            4 -> {/*EGenreErreurAcces.AucuneRessource*/}
            5 -> {/*EGenreErreurAcces.Connexion*/}
            6 -> {/*EGenreErreurAcces.BloqueeEleve*/}
            7 -> {/*EGenreErreurAcces.FonctionAccompagnant*/}
            8 -> throw AccessDeniedError()
            9 -> {
                if (data["AccesMessage"] != null) {
                    var error = data["AccesMessage"]!!.jsonObject["message"]?.jsonPrimitive?.content ?:  "(none)"

                    if (data["AccesMessage"]!!.jsonObject["titre"] != null)
                        error += "${data["AccesMessage"]!!.jsonObject["titre"]!!.jsonPrimitive.content} $error"

                    throw AuthenticateError(error)
                }
            }
        }

    return data
}